### PR TITLE
[SPARK-13697][PySpark]Fix the missing module name of TransformFunctionSerializer.loads

### PR DESCRIPTION
**Pull request by zsxwing**
## What changes were proposed in this pull request?

Set the function's module name to `__main__` if it's missing in `TransformFunctionSerializer.loads`.
## How was this patch tested?

Manually test in the shell.

Before this patch:

```
>>> from pyspark.streaming import StreamingContext
>>> from pyspark.streaming.util import TransformFunction
>>> ssc = StreamingContext(sc, 1)
>>> func = TransformFunction(sc, lambda x: x, sc.serializer)
>>> func.rdd_wrapper(lambda x: x)
TransformFunction(<function <lambda> at 0x106ac8b18>)
>>> bytes = bytearray(ssc._transformerSerializer.serializer.dumps((func.func, func.rdd_wrap_func, func.deserializers))) 
>>> func2 = ssc._transformerSerializer.loads(bytes)
>>> print(func2.func.__module__)
None
>>> print(func2.rdd_wrap_func.__module__)
None
>>> 
```

After this patch:

```
>>> from pyspark.streaming import StreamingContext
>>> from pyspark.streaming.util import TransformFunction
>>> ssc = StreamingContext(sc, 1)
>>> func = TransformFunction(sc, lambda x: x, sc.serializer)
>>> func.rdd_wrapper(lambda x: x)
TransformFunction(<function <lambda> at 0x108bf1b90>)
>>> bytes = bytearray(ssc._transformerSerializer.serializer.dumps((func.func, func.rdd_wrap_func, func.deserializers))) 
>>> func2 = ssc._transformerSerializer.loads(bytes)
>>> print(func2.func.__module__)
__main__
>>> print(func2.rdd_wrap_func.__module__)
__main__
>>> 
```

### CODE REVIEW

**Comment by zsxwing** [comment]
cc @davies

File: python/pyspark/streaming/util.py [hunk 1]
```diff
@@ -111,6 +111,10 @@ def loads(self, data):
         self.failure = None
         try:
             f, wrap_func, deserializers = self.serializer.loads(bytes(data))
+            if f.__module__ is None:
```
Comment (davies) [inline python/pyspark/streaming/util.py]: Could you fix this in CloudPicker?

Comment (zsxwing) [inline python/pyspark/streaming/util.py]: It just calls `pickle.loads(obj)`. So we cannot get the function object there.

Comment (davies) [inline python/pyspark/streaming/util.py]: checkout save_function_tuple() and _fill_function in cloudpickle.py

**Comment by davies** [comment]
```
diff --git a/python/pyspark/cloudpickle.py b/python/pyspark/cloudpickle.py
index 95b3abc..e56e22a 100644
--- a/python/pyspark/cloudpickle.py
+++ b/python/pyspark/cloudpickle.py
@@ -241,6 +241,7 @@ class CloudPickler(Pickler):
         save(f_globals)
         save(defaults)
         save(dct)
+        save(func.__module__)
         write(pickle.TUPLE)
         write(pickle.REDUCE)  # applies _fill_function on the tuple

@@ -698,13 +699,14 @@ def _genpartial(func, args, kwds):
     return partial(func, *args, **kwds)


-def _fill_function(func, globals, defaults, dict):
+def _fill_function(func, globals, defaults, dict, module):
     """ Fills in the rest of function data into the skeleton function object
         that were created via _make_skel_func().
          """
     func.__globals__.update(globals)
     func.__defaults__ = defaults
     func.__dict__ = dict
+    func.__module__ = module

     return func

```

**Comment by davies** [comment]
Please add a test case in SerializationTestCase

**Comment by zsxwing** [comment]
Updated and added a unit test. Thanks, @davies

**Comment by davies** [comment]
LGTM, merging this into master, thanks!

**Comment by davies** [comment]
Also merged into 1.4, 1.5, 1.6 branches.

**Pull request closed without merging**